File: source/src/main/java/org/rss/tools/mpl/render/html/FormattedHtmlRenderer.java
package org.rss.tools.mpl.render.html;

import java.io.StringReader;
import java.io.StringWriter;
import java.util.List;

import org.rss.tools.mpl.domain.Container;
import org.rss.tools.mpl.domain.Document;
import org.rss.tools.mpl.domain.Element;
import org.rss.tools.mpl.domain.Section;
import org.rss.tools.mpl.domain.element.Button;
import org.rss.tools.mpl.domain.element.Checkbox;
import org.rss.tools.mpl.domain.element.InputEmail;
import org.rss.tools.mpl.domain.element.InputSecret;
import org.rss.tools.mpl.domain.element.InputText;
import org.rss.tools.mpl.domain.element.Label;
import org.rss.tools.mpl.domain.element.LineHeader;
import org.rss.tools.mpl.domain.element.ListItem;
import org.rss.tools.mpl.domain.element.Radiobox;
import org.rss.tools.mpl.domain.element.Table;
import org.rss.tools.mpl.render.DocumentVisitor;
import org.rss.tools.mpl.render.RenderResponse;
import org.rss.tools.mpl.render.ResponseType;
import org.w3c.tidy.Tidy;

/**
 * Used to format the output created by other visitors. 
 * @author ricardo
 */
public class FormattedHtmlRenderer implements DocumentVisitor {

	private DocumentVisitor wrapped;

	public FormattedHtmlRenderer(DocumentVisitor wrapped) {
		super();
		if (wrapped == null) {
			throw new IllegalArgumentException("Source DocumentVisitor must be not null");
		}
		this.wrapped = wrapped;
	}

	public List<RenderResponse> getResult() {
		Tidy tidy = new Tidy();
		tidy.setIndentContent(true);
		tidy.setShowWarnings(false);

		StringWriter writer = new StringWriter();

		for (RenderResponse item : wrapped.getResult()) {
			if (item.getType().equals(ResponseType.StringData)) {
				tidy.parse(new StringReader((String) item.getData()), writer);
				item.setData(writer.toString());
			}
		}

		return wrapped.getResult();
	}

	public void startDocument(Document document) {
		wrapped.startDocument(document);
	}

	public void finishDocument(Document document) {
		wrapped.finishDocument(document);
	}

	public void visit(Document document) {
		wrapped.visit(document);
	}

	public void visit(Section element) {
		wrapped.visit(element);
	}

	public void visit(Button element) {
		wrapped.visit(element);
	}

	public void visit(Checkbox element) {
		wrapped.visit(element);
	}

	public void visit(InputEmail element) {
		wrapped.visit(element);
	}

	public void visit(InputSecret element) {
		wrapped.visit(element);
	}

	public void visit(InputText element) {
		wrapped.visit(element);
	}

	public void visit(Label element) {
		wrapped.visit(element);
	}

	public void visit(LineHeader element) {
		wrapped.visit(element);
	}

	public void visit(ListItem element) {
		wrapped.visit(element);
	}

	public void visit(Radiobox element) {
		wrapped.visit(element);
	}

	public void visit(Table element) {
		wrapped.visit(element);
	}

	public void visit(Container<?> element) {
		wrapped.visit(element);
	}

	public void visit(Element element) {
		wrapped.visit(element);
	}
}

File: source/src/main/kotlin/org/rss/tools/mpl/render/html/FormattedHtmlRenderer.kt
package org.rss.tools.mpl.render.html

import org.rss.tools.mpl.domain.Container
import org.rss.tools.mpl.domain.Document
import org.rss.tools.mpl.domain.Element
import org.rss.tools.mpl.domain.Section
import org.rss.tools.mpl.domain.element.*
import org.rss.tools.mpl.render.DocumentVisitor
import org.rss.tools.mpl.render.RenderResponse
import org.rss.tools.mpl.render.ResponseType
import org.w3c.tidy.Tidy

import java.io.StringReader
import java.io.StringWriter

/**
 * Used to format the output created by other visitors.
 * @author ricardo
 */
class FormattedHtmlRenderer(private val wrapped: DocumentVisitor) : DocumentVisitor {

    override//				System.out.println("Ignoring " + item.getData());
    val result: List<RenderResponse>
        get() {
            val tidy = Tidy()
            tidy.indentContent = true
            tidy.showWarnings = false

            val writer = StringWriter()

            for (item in wrapped.result) {
                if (item.type == ResponseType.StringData) {
                    tidy.parse(StringReader(item.data as String), writer)
                    item.data = writer.toString()
                } else {
                }
            }

            return wrapped.result
        }

    override fun startDocument(document: Document) {
        wrapped.startDocument(document)
    }

    override fun finishDocument(document: Document) {
        wrapped.finishDocument(document)
    }

    override fun visit(document: Document) {
        wrapped.visit(document)
    }

    override fun visit(element: Section) {
        wrapped.visit(element)
    }

    override fun visit(element: Button) {
        wrapped.visit(element)
    }

    override fun visit(element: Checkbox) {
        wrapped.visit(element)
    }

    override fun visit(element: InputEmail) {
        wrapped.visit(element)
    }

    override fun visit(element: InputSecret) {
        wrapped.visit(element)
    }

    override fun visit(element: InputText) {
        wrapped.visit(element)
    }

    override fun visit(element: Label) {
        wrapped.visit(element)
    }

    override fun visit(element: LineHeader) {
        wrapped.visit(element)
    }

    override fun visit(element: ListItem) {
        wrapped.visit(element)
    }

    override fun visit(element: Radiobox) {
        wrapped.visit(element)
    }

    override fun visit(element: Table) {
        wrapped.visit(element)
    }

    override fun visit(element: Container<*>) {
        wrapped.visit(element)
    }

    override fun visit(element: Element) {
        wrapped.visit(element)
    }
}
